libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -684,24 +701,43 @@ string ABIFunctions::abiEncodingFunctionCalldataArray(
 		_to.identifier() +
 		_options.toFunctionNameSuffix();
 	return createFunction(functionName, [&]() {
-		solUnimplementedAssert(fromArrayType.isByteArray(), "Only byte arrays can be encoded from calldata currently.");
-		// TODO if this is not a byte array, we might just copy byte-by-byte anyway,
-		// because the encoding is position-independent, but we have to check that.
-		Whiskers templ(R"(
-			// <readableTypeNameFrom> -> <readableTypeNameTo>
-			function <functionName>(start, length, pos) -> end {
-				pos := <storeLength>(pos, length)
-				<copyFun>(start, pos, length)
-				end := add(pos, <lengthPadded>)
-			}
-		)");
-		templ("storeLength", arrayStoreLengthForEncodingFunction(toArrayType, _options));
-		templ("functionName", functionName);
-		templ("readableTypeNameFrom", _from.toString(true));
-		templ("readableTypeNameTo", _to.toString(true));
-		templ("copyFun", m_utils.copyToMemoryFunction(true));
-		templ("lengthPadded", _options.padded ? m_utils.roundUpFunction() + "(length)" : "length");
-		return templ.render();
+		bool needsPadding = _options.padded && fromArrayType.isByteArray();
+		if (fromArrayType.isDynamicallySized())
+		{
+			Whiskers templ(R"(
+				// <readableTypeNameFrom> -> <readableTypeNameTo>
+				function <functionName>(start, length, pos) -> end {
+					pos := <storeLength>(pos, length)
+					let byteLength := mul(length, <stride>)
```

Comment:
Can we be sure that this does not overflow?

Reply:
Added a check.